File: rust/7.rs
// Project Euler problem #7 in Rust by antekone.
// http://anadoxin.org/blog
//
// usage:
//
// $ rustc 7.rs
// $ ./7 | grep 10001
//
// The answer will be printed on the screen ;)

use std::num::Float;

// Sieve of Erastothenes.

fn e_sieve(n: usize) {
    let mut A: Vec<bool> = Vec::new();
    A.resize(n, true);

    let sqrtn: usize = (n as f32).sqrt() as usize;
    for i in (2us..sqrtn) {
        let mut j = i * i;

        if(! A[i]) {
            continue;
        }

        while(j < n) {
            A[j] = false;
            j += i;
        }
    }

    let mut cnt = 1us;
    for i in (2us..n) {
        if(A[i]) {
            println!("idx={} prime={}", cnt, i);
            cnt += 1;
        }
    }
}

fn main() {
    e_sieve(211000);
}

File: c++/7.cc
// Project Euler #7 in C++ by antekone
// http://anadoxin.org/blog
//
// Usage:
//
// $ g++ 7.cc -o 7
// $ ./7 | grep 10001

#include <vector>
#include <iostream>
#include <math.h>

// Sieve of Eratosthenes.

void e_sieve(size_t n) {
	std::vector<bool> A;
	A.resize(n);
	std::fill(A.begin(), A.end(), true);

	size_t sqrtn = sqrt(n);
	for(size_t i = 2; i < sqrtn; ++i) {
		size_t j = i * i;
		if(! A[i])
			continue;

		while(j < n) {
			A[j] = false;
			j += i;
		}
	}

	size_t cnt = 1;
	for(size_t i = 2; i < n; ++i) {
		if(A[i]) {
			std::cout << "idx=" << cnt << " prime=" << i << "\n";
			++cnt;
		}
	}
}

int main() {
	e_sieve(211000);
	return 0;
}
